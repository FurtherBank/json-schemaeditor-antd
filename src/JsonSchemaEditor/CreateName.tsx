import { CheckOutlined, CloseOutlined, PlusOutlined } from "@ant-design/icons"
import AutoComplete from "antd/lib/auto-complete"
import Button from "antd/lib/button"
import Input from "antd/lib/input"
import message from "antd/lib/message"
import _ from "lodash"
import React, { useState } from "react"
import { SchemaCache } from "."
import { FieldProps } from "./Field"
import { FatherInfo } from "./FieldList"
import { getDefaultValue } from "./FieldOptions"
import { PropertyInfo } from "./reducer"
import { addRef, concatAccess, findKeyRefs, getValueByPattern, iterToArray, matchKeys } from "./utils"

interface CreateNameProps {
  fatherInfo: FatherInfo
  fieldProps: FieldProps
  fieldCache: SchemaCache
  style?: React.CSSProperties
}

const CreateName = (props: CreateNameProps) => {
  const { fatherInfo, fieldProps, style, fieldCache } = props
  const { data, doAction, route, field} = fieldProps
  const { valueEntry, propertyCache, itemCache, valueSchemaMap } = fieldCache
  const access = concatAccess(route, field)

  const [editing, setEditing] = useState(false)
  const [name, setName] = useState("")

  const propertyCacheValue =
    fatherInfo.type === "object" && valueEntry
      ? propertyCache.get(valueEntry)
      : undefined

  let autoCompleteOptions: any[] = []
  if (propertyCacheValue) {
    const { props, patternProps } = propertyCacheValue
    const optionStrings = props ? Object.keys(props).filter((key) => {
      return data[key] === undefined
    }) : []
    autoCompleteOptions = optionStrings.map((key) => {
      return { value: key }
    })
  }
  const handleClick = (e: React.SyntheticEvent) => {
    if (fatherInfo.type === "object") {
      setEditing(!editing)
    } else {
      // 数组新变量创建。注意如果使用已有变量直接创建时不要忘记深拷贝！
      const itemCacheValue = valueEntry ? itemCache.get(valueEntry) : undefined
      const nowLength = data.length
      if (itemCacheValue) {
        const { itemLength } = itemCacheValue
        const schemaRef = itemLength !== undefined ? nowLength < itemLength ? addRef(valueEntry, 'items', nowLength.toString()) : addRef(valueEntry, 'additionalItems') : addRef(valueEntry, 'items')
        const defaultValue = getDefaultValue(fieldCache, schemaRef)
        doAction!('create', access, nowLength.toString(), defaultValue)
      } else {
        doAction!('create', access, nowLength.toString(), _.cloneDeep(data[data.length-1] ?? null))
      }
    }
  }
  const handleNameChange = (value: string) => {
    setName(value)
  }
  const handleCreate = (e: React.SyntheticEvent) => {
    let newValueEntry = undefined
    if (data[name] !== undefined) {
      message.error(`字段 ${name} 已经存在！`)
      return
    } else {
      if (propertyCacheValue) {
        const { props, patternProps, additional } = propertyCacheValue
        const info = props[name] ?? getValueByPattern(patternProps, name) ?? additional
        if (!info) {
          message.error(
            `${name} 不匹配 properties 中的名称或 patternProperties 中的正则式`
          )
          return
        }
        const {ref, shortable} = info
        if (ref) {
          newValueEntry = ref
        } else {
          newValueEntry = findKeyRefs(valueSchemaMap!, 'additionalProperties') as string
        }
      }
    }
    doAction!('create', access, name, getDefaultValue(fieldCache, newValueEntry))
    setEditing(false)
  }
  return editing ? (
    <div style={{display: 'flex'}}>
      <AutoComplete
        options={autoCompleteOptions}
        filterOption={(inputValue, option) =>
          option!.value.toUpperCase().indexOf(inputValue.toUpperCase()) !== -1
        }
        
        value={name}
        onChange={handleNameChange}
        style={{flex: '1'}}
      >
        <Input size="small" addonBefore={"新属性名称"} onPressEnter={handleCreate}/>
      </AutoComplete>

      <Button size="small" shape="circle" onClick={handleCreate} style={{margin: '0 6px'}}>
        <CheckOutlined />
      </Button>
      <Button size="small" shape="circle" onClick={handleClick}>
        <CloseOutlined />
      </Button>
    </div>
  ) : (
    <Button
      type="dashed"
      icon={<PlusOutlined />}
      size="small"
      block
      onClick={handleClick}
      style={style}
    >
      {fatherInfo.type === "object" ? "Property" : "Item"}
    </Button>
  )
}

export default CreateName
