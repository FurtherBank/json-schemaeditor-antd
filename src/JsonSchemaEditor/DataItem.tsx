import React, { PropsWithChildren } from 'react';
import { TSelectableItemProps, createSelectable } from 'react-selectable-fast';

export type DataItemProps = {
  id: number;
};

const DataItem = createSelectable<DataItemProps>(
  (props: TSelectableItemProps & PropsWithChildren<DataItemProps>) => {
    const { selectableRef, isSelected, isSelecting, children } = props;

    const classNames = [
      'ant-select-item ant-select-item-option ant-select-item-option-active',
      false,
      isSelecting && 'ant-select-item-option-selected',
      isSelected && 'ant-select-item-option-selected',
    ]
      .filter(Boolean)
      .join(' ');

    return (
      <div ref={selectableRef} className={classNames}>
        {children}
      </div>
    );
  },
);

export default DataItem;
