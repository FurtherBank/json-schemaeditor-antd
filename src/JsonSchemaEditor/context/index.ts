import { getRefSchemaMap, pathGet } from '../utils'
import { ofSchemaCache, setOfInfo } from './ofInfo'
import { MergedSchema, mergeSchemaMap } from './mergeSchema'
import { doAction, CpuEditorState, CpuEditorAction } from '../definition/reducer'
import { AnyAction, Dispatch, Store } from 'redux'
import { StateWithHistory } from 'redux-undo'
import { JSONSchema } from '../type/Schema'
import { IComponentMap, IViewsMap } from '../type/Components'
import { antdComponentMap, antdViewsMap } from '../components/antd'
import Field from '../Field'
import { ComponentType } from 'react'

export interface SchemaArrayRefInfo {
  ref: string
  length: number
}

/**
 * 整个`editor`的上下文，包含了：
 * - redux store
 * - `schema`的各个`ref`对应的模式性质信息
 * 在编辑器`schema`变更时重新初始化
 */
export default class CpuEditorContext {
  Field = Field
  rootSchema: JSONSchema | false
  mergedSchemaMap: Map<string, MergedSchema | false>
  /**
   * schemaInfo 是 schema 信息的需要对子层进行归纳的部分。
   * 将 schemaInfo 和 mergedSchema 分开，是为了编辑器可以按需创建这些信息。
   */
  ofInfoMap: Map<string, ofSchemaCache | null>
  /**
   * 已加载资源的缓存列表，通过`url`为键索引。
   */
  resourceMap: Map<string, any>

  dispatch: Dispatch<AnyAction>

  executeAction: (type: string, route?: string[], field?: string, value?: any) => CpuEditorAction
  private actionCreator: (type: string, route?: string[], field?: string, value?: any) => CpuEditorAction

  constructor(
    rootSchema: false | JSONSchema,
    public id: string | undefined,
    public store: Store<StateWithHistory<CpuEditorState>, AnyAction>,
    public setDrawer: (route: string[], field: string | undefined) => void,
    public readonly componentMap: IComponentMap = antdComponentMap,
    /**
     * 自定义 view 的组件列表，通过`view.type`索引到 componentMap；
     *
     * 其中 componentMap 的组件都是非必须的。
     *
     * 如果在 schema 中限定了`view.type`，但没有在对应的 componentMap 找到组件，将使用对应的默认组件代替。
     */
    public readonly viewsMap: Record<string, IViewsMap> = antdViewsMap
  ) {
    this.rootSchema = rootSchema
    this.mergedSchemaMap = new Map()
    this.ofInfoMap = new Map()
    this.resourceMap = new Map()
    this.dispatch = store.dispatch
    this.actionCreator = doAction
    this.executeAction = (type: string, route?: string[], field?: string, value?: any) => {
      const action = this.actionCreator(type, route, field, value)
      this.dispatch(action)
      return action
    }
  }

  /**
   * 得到对应 ref 的合并的模式。
   * 如果 ref 不存在则返回空对象
   * @param ref
   * @returns
   */
  getMergedSchema(ref: string | undefined) {
    if (this.rootSchema === false || !ref) return false
    return this.mergedSchemaMap.has(ref) ? this.mergedSchemaMap.get(ref)! : this.makeMergedSchema(ref)
  }

  /**
   * 得到 ref 对应模式的 of 信息。
   *
   * @param ref
   * @returns
   */
  getOfInfo(ref: string | undefined) {
    if (this.rootSchema === false || !ref) return null
    return this.ofInfoMap.has(ref) ? this.ofInfoMap.get(ref)! : setOfInfo(this, ref, this.rootSchema)
  }

  /**
   * 从 componentMap 找到对应 role path 的组件。
   *
   * 和`getComponent`的区别是，若未找到，会 fallback 为对应的`string`编辑组件。
   * @param view 组件的`viewType`(必须经过验证)
   * @param rolePath 组件角色路径
   * @returns
   */
  getComponent(view: string | null = null, rolePath: string | string[]): ComponentType<any> {
    const componentPath = typeof rolePath === 'string' ? [rolePath] : rolePath
    if (view) {
      const result = pathGet(this.viewsMap[view] || {}, componentPath)

      if (result) return result
    }
    const result = pathGet(this.componentMap, componentPath)
    if (result) {
      return result
    } else {
      throw new Error(`The component of ${componentPath} not exist in component map`)
    }
  }

  /**
   * 从 componentMap 找到对应 string format 的组件。
   *
   * 和`getComponent`的区别是，若未找到，会 fallback 为对应的`string`编辑组件。
   * @param view 组件的`viewType`(必须经过验证)
   * @param format
   * @returns
   */
  getFormatComponent(view: string | null = null, format: string): ComponentType<any> {
    const componentPath = ['format', format]
    if (view) {
      const result = pathGet(this.viewsMap, componentPath)
      if (result) return result
    }
    const result = pathGet(this.componentMap, componentPath)
    if (result) {
      return result
    } else {
      return this.getComponent(view, ['edition', 'string'])
    }
  }

  private makeMergedSchema(ref: string) {
    if (typeof this.rootSchema === 'boolean') return false
    const map = getRefSchemaMap(ref, this.rootSchema)
    const mergedSchema = mergeSchemaMap(this, map)
    this.mergedSchemaMap.set(ref, mergedSchema)
    return mergedSchema
  }
}
