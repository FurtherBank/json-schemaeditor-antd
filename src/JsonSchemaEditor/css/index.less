
.selectable-selectbox {
  z-index: 9000;
  position: absolute;
  cursor: default;
  background: none;
  border: 1px dashed grey;
}

.selectable-div, .selected-div {
  width: 180px;
  height: 28px;
  line-height: 28px;
  padding: 0 3px;
  box-sizing: border-box;
  user-select:none;
}

.selectable-div {
  background-color: #ffffff;
  :hover {
    background-color: #eeeeee;
  }
}

.selected-div {
  background-color: #dddddd;
  :hover {
    background-color: #6699bb;
  }
}