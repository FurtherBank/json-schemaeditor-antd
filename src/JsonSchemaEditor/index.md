---
nav:
  title: 预览
---

## JSON Schema Editor

<code src="./demos/App.tsx"></code>

好家伙，有谁能想到用 Markdown 写代码呢？是哪位工程师思路如此清奇？  
More skills for writing demo: https://d.umijs.org/guide/basic#write-component-demo

第一个问题：`@umijs/plugin-antd`对 dumi 的 md 编译的 jsx 不过 babel，没有按需引入优化。

## demo2

我是谁？
