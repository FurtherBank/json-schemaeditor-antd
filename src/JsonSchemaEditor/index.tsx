import React, {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { createStore } from 'redux';
import { Provider } from 'react-redux';
import Field from './Field';
import {
  ajvInstance,
  itemSchemaCache,
  ofSchemaCache,
  propertySchemaCache,
  reducer,
} from './reducer';
import FieldDrawer from './FieldDrawer';
import { Alert } from 'antd';

import './css/index.less';
import _ from 'lodash';
import { JSONSchema6 } from 'json-schema';

interface EditorProps {
  onChange?: (data: any) => void | null;
  data?: any;
  schema: JSONSchema6 | true;
}
export const CacheContext = React.createContext({
  ofCache: new Map(),
  propertyCache: new Map(),
  itemCache: new Map(),
  rootSchema: {},
} as ContextContent);

export interface ContextContent {
  ofCache: Map<string, ofSchemaCache | null>;
  propertyCache: Map<string, propertySchemaCache | null>;
  itemCache: Map<string, itemSchemaCache | null>;
  rootSchema: JSONSchema6;
}

export interface SchemaCache extends ContextContent {
  entrySchemaMap: Map<string, boolean | JSONSchema6>;
  valueEntry: string | undefined;
  valueSchemaMap: Map<string, boolean | JSONSchema6>;
}

const emptyArray: never[] = [];

const Editor = (props: EditorProps, ref: React.ForwardedRef<any>) => {
  const { schema, data, onChange } = props;

  // useMemo 编译 schema
  const validate = useMemo(() => {
    let validate = undefined;
    let schemaErrors = null;

    console.time('compile schema');
    try {
      validate = ajvInstance.compile(schema);
    } catch (error) {
      schemaErrors = error;
    }
    console.timeEnd('compile schema');

    return validate ? validate : schemaErrors;
  }, [schema]) as Function | any;

  // useMemo 初始化 store
  const store = useMemo(() => {
    const initialState = {
      data: data,
      dataErrors: [],
      validate: typeof validate === 'function' ? validate : undefined,
    };

    const store = createStore(reducer, {
      past: [],
      present: initialState,
      future: [],
    });

    const change = () => {
      const changedData = store.getState().present.data;
      if (onChange && typeof onChange == 'function') {
        onChange(changedData);
      }
    };
    store.subscribe(change);

    return store;
  }, [schema]);

  // 暴露一下 api
  useImperativeHandle(
    ref,
    () => {
      return store;
    },
    [store],
  );

  const caches = useMemo(() => {
    console.log('caches变化');
    return {
      ofCache: new Map(),
      propertyCache: new Map(),
      itemCache: new Map(),
      rootSchema: validate instanceof Function ? (typeof schema !== 'boolean' ? schema : {}) : {},
    };
  }, [schema]);

  // 如果 data 更新来自外部，通过 setData 与 store 同步
  const presentData = store.getState().present.data;
  if (data !== presentData) {
    console.log('检测到外部更新：', data, presentData);
    store.dispatch({
      type: 'setData',
      value: data,
    });
  }

  // 详细抽屉功能
  const drawerRef = useRef(null) as React.RefObject<any>;
  const setDrawer = useCallback(
    (...args: any[]) => {
      console.log('setDrawer', drawerRef.current);
      if (drawerRef.current) drawerRef.current.setDrawer(...args);
    },
    [drawerRef],
  );

  return (
    <Provider store={store}>
      {validate instanceof Function ? null : (
        <Alert message="Error" description={validate.toString()} type="error" showIcon />
      )}
      <CacheContext.Provider value={caches}>
        <Field route={emptyArray} field={null} schemaEntry="#" setDrawer={setDrawer} />
        <FieldDrawer ref={drawerRef} />
      </CacheContext.Provider>
    </Provider>
  );
};

export default React.memo(forwardRef(Editor));
