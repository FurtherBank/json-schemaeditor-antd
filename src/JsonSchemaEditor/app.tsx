import React, { useMemo, useRef, useState, useEffect, useCallback } from "react"
import _, { cloneDeep } from "lodash"

import { Button, Card, PageHeader, message, Modal, Select } from "antd"  // 用 antd 封装demo
import MonacoEditor from "react-monaco-editor/lib/editor"

import JsonSchemaEditor from "."  // 不知道为什么，直接写"json-schemaeditor-antd"会报错！有什么解决方案？

// json 配置
import general from "./json-example/general.json"
import $general from "./json-example/$schema.general.json"
import $default from "./json-example/$schema.default.json"
import Default from "./json-example/default.json" // 注意 小写 default 保留字

import $dataFacility from "./json-example/$schema.dataFacility.json"
import dataFacility from "./json-example/dataFacility.json"

import $dataTechTree from "./json-example/$schema.dataTechTree.json"
import dataTechTree from "./json-example/dataTechTree.json"

import $items from "./json-example/$schema.items.json"
import items from "./json-example/items.json"

import $basic from "./json-example/$schema.basic.json"
import basic from "./json-example/basic.json"

import $simple from "./json-example/$schema.simple.json"
import simple from "./json-example/simple.json"

import meta from "./json-example/$meta.json"

const examples = [
  ['基础', basic, $basic],
  ['一系列测试', general, $general],
  ['小型示例', Default, $default],
  ['简单示例', simple, $simple],
  ['模式编辑', $default, meta],
  ['元模式自编辑', meta, meta],
  ['《星际探索者》设施配置示例', dataFacility, $dataFacility],
  ['《星际探索者》设施配置模式编辑', $dataFacility, meta],
  ['《星际探索者》科技树示例', dataTechTree, $dataTechTree],
  ['《星际探索者》科技树配置模式编辑', $dataTechTree, meta],
  ['RMMZ 物品数据示例', items, $items],
]

/**
 * 节流：设置 func 在 interval 时间内只执行一次。  
 * 函数会随依赖改变，且保证组件卸载等情况下，函数可以跟进执行最后一次。
 * @param func 
 * @param interval 
 * @param deps 
 * @returns 
 */
const useCooldown = <T extends (...args: any[]) => void>(
  func: T,
  interval = 1,
  deps: React.DependencyList
): T => {
  let cd = 0 
  let newArgs = undefined as undefined | any[]
  const rf = (...args: any[]) => {
    if (cd === 0) {
      func(...args)
      cd = setTimeout(() => {
        if (newArgs) func(...newArgs)
        newArgs = undefined
        cd = 0
      }, interval)
    } else {
      newArgs = args
    }
  }
  // 在组件卸载时，清除计时器
  useEffect(() => {
    // 不加载，只清除
    return () => {
      if (cd) {
        clearTimeout(cd)
        if (newArgs) func(...newArgs)
      }
    }
  }, deps)

  return useCallback(rf as T, deps)
}

// 接下来是示例选择功能的定义
const options = examples.map((example, i) => {
  return { label: example[0], value: i.toString() }
}) as unknown as { label: string; value: string }[]

const ModalSelect = (props: { cb: (data: any, schema: any) => void, cancelCb: () => void, visible: boolean}) => {
  const {cb, cancelCb, visible} = props
  const [item, setItem] = useState(0)

  return (
    <Modal
      title="选择示例"
      okText="选择"
      cancelText="取消"
      onOk={() => {
        const data = cloneDeep(examples[item][1]), schema = cloneDeep(examples[item][2])
        cb(data, schema)
      }}
      onCancel={() => {
        cancelCb()
      }}
      visible={visible}
    >
      <Select
        showSearch
        placeholder="选择示例"
        optionFilterProp="label"
        onChange={(value) => {
          setItem(parseInt(value))
        }}
        filterOption={(input, option) => {
          return !!option && option.label.toLowerCase().indexOf(input.toLowerCase()) >= 0
        }}
        defaultValue={'0'}
        options={options}
        style={{width: "100%"}}
      />
    </Modal>
  )
}

// 从 localhost 读取 json
const loadLocalJson = (key: string) => {
  try {
    const value = localStorage.getItem(key)
    return value ? JSON.parse(value) : undefined
  } catch (error) {
    return undefined
  }
}

// app 最终定义
export default () => {
  const [mode, setMode] = useState(0)

  const [data, setData] = useState(() => loadLocalJson("data") ?? cloneDeep(examples[0][1]))
  const [schema, setSchema] = useState(() => loadLocalJson("schema") ?? cloneDeep(examples[0][2]))

  const dataEditor = useRef<MonacoEditor>(null)
  const schemaEditor = useRef<MonacoEditor>(null)

  const [isModalVisible, setIsModalVisible] = useState(false)

  const showModal = () => {
    setIsModalVisible(true)
  }

  const handleCancel = () => {
    setIsModalVisible(false)
  }

  // 自动保存 
  // 注：开发模式会渲染两次，这里就会保存两遍。生产模式没事
  const save = useCooldown(
    (data, schema) => {
      try {
        localStorage.setItem("data", JSON.stringify(data))
        localStorage.setItem("schema", JSON.stringify(schema))
      } catch (error) {
        console.error(error)
      }
    },
    1000,
    []
  )

  useMemo(() => {
    save(data, schema)
  }, [data, schema])

  const changeData = (value: any) => {
    setData(typeof value === 'object' ? value : value)
  }

  const changeExample = (data: any, schema: any) => {
    setData(data)
    setSchema(schema)
    setIsModalVisible(false)
  }

  const changeMode = () => {
    if (mode) {
      try {
        // get code
        const dataCode = dataEditor.current?.editor?.getValue() ?? "{}"
        const schemaCode = schemaEditor.current?.editor?.getValue() ?? "{}"
        // parse json code
        const newData = JSON.parse(dataCode),
          newSchema = JSON.parse(schemaCode)
        setData(newData)
        setSchema(newSchema)
        setMode(0)
      } catch (error) {
        message.error(`JSON 数据 或者 JSON JSONSchema6 解析出错，请检查代码编辑器中的数据是否正确。`)
      }
    } else {
      setMode(1)
    }
  }

  return (
    <div style={{ height: "100vh", overflow: "auto", display: "flex", flexDirection: "column" }}>
      <PageHeader
        title="JSON Editor"
        className="site-page-header"
        subTitle="By FurtherBank"
        extra={[
          <Button key="2" onClick={changeMode}>
            {mode ? "返回编辑器" : "源代码编辑"}
          </Button>,
          <Button key="1" type="primary" onClick={showModal}>
            加载示例
          </Button>,
        ]}
      ></PageHeader>
      {mode ? (
        <div style={{ display: "flex", overflow: "auto", justifyContent: "stretch", flex: 1 }}>
          <Card title="data" style={{ flex: 1, display: "flex", flexDirection: "column" }} bodyStyle={{ flex: 1 }}>
            <MonacoEditor
              language="json"
              theme="vs"
              value={JSON.stringify(data, null, 2)}
              ref={dataEditor}
              options={{
                automaticLayout: true,
                formatOnPaste: true,
                formatOnType: true,
                renderLineHighlight: "none",
                autoClosingOvertype: "always",
                cursorStyle: "line",
                quickSuggestions: false,
                scrollBeyondLastLine: false,
                snippetSuggestions: "none",
                minimap: {
                  enabled: true,
                },
              }}
            />
          </Card>
          <Card title="schema" style={{ flex: 1, display: "flex", flexDirection: "column" }} bodyStyle={{ flex: 1 }}>
            <MonacoEditor
              language="json"
              theme="vs"
              value={JSON.stringify(schema, null, 2)}
              ref={schemaEditor}
              options={{
                automaticLayout: true,
                formatOnPaste: true,
                formatOnType: true,
                renderLineHighlight: "none",
                autoClosingOvertype: "always",
                cursorStyle: "line",
                quickSuggestions: false,
                scrollBeyondLastLine: false,
                snippetSuggestions: "none",
                minimap: {
                  enabled: true,
                },
              }}
            />
          </Card>
        </div>
      ) : (
        <JsonSchemaEditor data={data} schema={schema as any} onChange={changeData} />
      )}
      <ModalSelect cb={changeExample} cancelCb={handleCancel} visible={isModalVisible}/>
    </div>
  )
}

