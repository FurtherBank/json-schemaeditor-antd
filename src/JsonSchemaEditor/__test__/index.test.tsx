import '@testing-library/jest-dom';
import React from 'react';
import { render, screen } from '@testing-library/react';
import JsonSchemaEditor from '..';

describe('<JsonSchemaEditor />', () => {
  it('should not crash', () => {
    const schema = {
      type: 'string',
      format: 'multiline',
    };
    render(<JsonSchemaEditor data={'hello world'} schema={schema as any} />);
    expect(screen.queryByText('hello world')).toBeInTheDocument();
  });
});
