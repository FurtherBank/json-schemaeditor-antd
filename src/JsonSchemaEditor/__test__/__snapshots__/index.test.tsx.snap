// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<JsonSchemaEditor /> should not crash 1`] = `
<DocumentFragment>
  <div
    class="ant-card ant-card-bordered ant-card-small"
  >
    <div
      class="ant-card-head"
    >
      <div
        class="ant-card-head-wrapper"
      >
        <div
          class="ant-card-head-title"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
            >
              <span
                style="width: 100px;"
                title=" "
              >
                 
              </span>
            </div>
          </div>
        </div>
        <div
          class="ant-card-extra"
        >
          <div
            class="ant-space ant-space-horizontal ant-space-align-center"
          >
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <input
                class="ant-input ant-input-sm"
                style="flex: 1;"
                type="text"
                value="A basic string"
              />
            </div>
            <div
              class="ant-space-item"
              style="margin-right: 8px;"
            >
              <button
                class="ant-btn ant-btn-circle ant-btn-default ant-btn-sm ant-btn-icon-only"
                type="button"
              >
                <span
                  aria-label="undo"
                  class="anticon anticon-undo"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="undo"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M511.4 124C290.5 124.3 112 303 112 523.9c0 128 60.2 242 153.8 315.2l-37.5 48c-4.1 5.3-.3 13 6.3 12.9l167-.8c5.2 0 9-4.9 7.7-9.9L369.8 727a8 8 0 00-14.1-3L315 776.1c-10.2-8-20-16.7-29.3-26a318.64 318.64 0 01-68.6-101.7C200.4 609 192 567.1 192 523.9s8.4-85.1 25.1-124.5c16.1-38.1 39.2-72.3 68.6-101.7 29.4-29.4 63.6-52.5 101.7-68.6C426.9 212.4 468.8 204 512 204s85.1 8.4 124.5 25.1c38.1 16.1 72.3 39.2 101.7 68.6 29.4 29.4 52.5 63.6 68.6 101.7 16.7 39.4 25.1 81.3 25.1 124.5s-8.4 85.1-25.1 124.5a318.64 318.64 0 01-68.6 101.7c-7.5 7.5-15.3 14.5-23.4 21.2a7.93 7.93 0 00-1.2 11.1l39.4 50.5c2.8 3.5 7.9 4.1 11.4 1.3C854.5 760.8 912 649.1 912 523.9c0-221.1-179.4-400.2-400.6-399.9z"
                    />
                  </svg>
                </span>
              </button>
            </div>
            <div
              class="ant-space-item"
            >
              <button
                class="ant-btn ant-btn-circle ant-btn-default ant-btn-sm ant-btn-icon-only"
                type="button"
              >
                <span
                  aria-label="redo"
                  class="anticon anticon-redo"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="redo"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M758.2 839.1C851.8 765.9 912 651.9 912 523.9 912 303 733.5 124.3 512.6 124 291.4 123.7 112 302.8 112 523.9c0 125.2 57.5 236.9 147.6 310.2 3.5 2.8 8.6 2.2 11.4-1.3l39.4-50.5c2.7-3.4 2.1-8.3-1.2-11.1-8.1-6.6-15.9-13.7-23.4-21.2a318.64 318.64 0 01-68.6-101.7C200.4 609 192 567.1 192 523.9s8.4-85.1 25.1-124.5c16.1-38.1 39.2-72.3 68.6-101.7 29.4-29.4 63.6-52.5 101.7-68.6C426.9 212.4 468.8 204 512 204s85.1 8.4 124.5 25.1c38.1 16.1 72.3 39.2 101.7 68.6 29.4 29.4 52.5 63.6 68.6 101.7 16.7 39.4 25.1 81.3 25.1 124.5s-8.4 85.1-25.1 124.5a318.64 318.64 0 01-68.6 101.7c-9.3 9.3-19.1 18-29.3 26L668.2 724a8 8 0 00-14.1 3l-39.6 162.2c-1.2 5 2.6 9.9 7.7 9.9l167 .8c6.7 0 10.5-7.7 6.3-12.9l-37.3-47.9z"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="ant-card-body"
      style="display: none;"
    />
  </div>
</DocumentFragment>
`;
